README.md
rospilot
========

1. Install the dependencies: `rosdep install rospilot`
2. Compile: `catkin_make`
3. Configure PostGIS, mapping server, and other services: `rosrun rospilot first_time_setup.sh`
4. Start rospilot: `roslaunch rospilot rospilot.launch`
5. The UI will now be running at http://localhost:8085

I am providing code in this repository to you under an open source license. Because this is my personal repository, the license you receive to my code is from me and not from my employer (Facebook).


Running linter
==============
`catkin_make roslint_rospilot`


Running tests
=============
`catkin_make test`


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package rospilot
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.6.1 (2021-02-21)
------------------
* Fix systemd service to use noetic
* Remove map server
  Tilestache is no longer supported on modern Ubuntu
* Remove firmware directory
* Switch to Github Actions for CI
* Remove calls to deprecated ffmpeg functions
* Contributors: Christopher Berner

1.6.0 (2020-09-06)
------------------
* Add support for Noetic
* Update to Python3
* Upgrade jquery to fix CVE
* Contributors: Christopher Berner

1.5.6 (2019-03-30)
------------------
* Avoid thread exhaustion in h264 server
  Previously, connections didn't timeout, so a badly behaved client
  could exhaust all the threads on the host
* Contributors: Christopher Berner

1.5.5 (2019-03-29)
------------------
* Fix web_ui.py serving of nodejs dependencies
* Upgrade Bootstrap to fix CVEs
* Contributors: Christopher Berner

1.5.4 (2019-02-17)
------------------
* Update vendored NPM deps
* Remove hard npm dependency
  npm deps must now be vendored
* Remove Highcharts and accelerometer graph page
* Added missing wget dependency to package.xml that was found in first_time_setup.sh.
* Improve cold start H264 streaming to new clients
* Refactor H264 streaming to allow Android app support
* Contributors: Anders Fischer, Christopher Berner

1.5.3 (2018-10-08)
------------------
* Fix dnsmasq dependency on wlan device
* Contributors: Christopher Berner

1.5.2 (2018-09-22)
------------------
* Fix compiler warnings
* Fix SPS & PPS ordering when using MFC encoder
  SPS should always be before PPS, as PPS references the SPS packet
* Remove unnecessary C++11 compiler flag
* Fix undefined behavior in AVPacket buffer management
  Use the provided av_new_packet() and av_packet_unref()
  functions instead of manual management of .data field
* Fix undefined behavior in memory deallocation
  This buffer is an array, and therefore must be deallocated with delete[]
* Contributors: Christopher Berner

1.5.1 (2018-09-03)
------------------
* Fix wlan configuration on Bionic
* Fix Systemd service to work on Melodic
* Contributors: Christopher Berner

1.5.0 (2018-08-04)
------------------
* Upgrade to Angular 5.2.11
* Upgrade Highcharts
* Set both pts and dts in video recorder
* Properly set time_base on video recorder codec
* Remove usage of deprecated encoding/decoding functions
* Remove usage of deprecated AVPicture
* Fix AVStream::codec deprecation warnings
* Add error handling for call to avformat_write_header()
* Cleanup VideoRecorder::stop()
* Switch Travis to build melodic
* Update ffmpeg usage to compile against newest version
* Update environment variable to PGPASSWORD
* Switch to Melodic-friendly dependencies
* Hack: monkey patch Mapnik to fix Tilestache dependency
* Contributors: Christopher Berner

1.4.1 (2018-03-10)
------------------
* Switch to HTTPS for openstreetmap.org
* Workaround camera drivers that don't set keyframe flag
* Add support for reading raw YUV from camera
* Refactor pixel format handling
* Contributors: Christopher Berner

1.4.0 (2017-11-19)
------------------
* Add "follow me" mode
* Add support for direct h264 capture from camera
* Improve automatic camera selection
* Add vibration data to display
  Also includes accelerometer clipping counts
* Add support for battery voltage monitoring
* Enable pre-release tests
* Add Travis integration
* Refactor handling of NPM dependencies
* Add native Systemd service
* Set dns search domain for better compatibility
* Update wifi setup script with proper dnsmasq dependency
* Contributors: Christopher Berner

1.3.8 (2017-09-13)
------------------
* Remove dependency on nodejs-legacy
  Also remove usage of vendored npm version
* Contributors: Christopher Berner

1.3.7 (2017-06-10)
------------------
* Update for newest NPM
* Contributors: Christopher Berner

1.3.6 (2017-05-29)
------------------
* Add support for Pixhawk2.1
* Contributors: Christopher Berner

1.3.5 (2017-03-12)
------------------
* Add missing include needed on Debian
* Contributors: Christopher Berner

1.3.4 (2017-03-08)
------------------
* Rescale PTS properly
* Fix usage of deprecated API
* Fix usage of invalid iterator
* Add more logging to camera node
* Fix data structure corruption due to race
* Contributors: Christopher Berner

1.3.3 (2016-08-24)
------------------
* Fix build on Debian
* Contributors: Christopher Berner

1.3.2 (2016-08-24)
------------------
* Support building on Debian
* Contributors: Christopher Berner

1.3.1 (2016-08-21)
------------------
* Fix build on Wily
* Contributors: Christopher Berner

1.3.0 (2016-08-20)
------------------
* Initialize source frame data structure
* Update init script to Kinetic
* Set CherryPy to production instead of manually configuring it
* Use npm for build
* Use more canonical license string
* Fix scoping in set waypoint callback
* Improve mavlink wait logic to be interruptable
* Fix retry logic for serial mavlink
* Migrate to Angular 2.0
* Add timeout to waypoint fetching
  Previously, if a message was lost waypoints would no longer be fetched
  and new ones could not be set
* Contributors: Christopher Berner

1.2.0 (2016-05-14)
------------------
* Add people detector using OpenCV
* Contributors: Christopher Berner

1.1.1 (2016-01-31)
------------------
* Copy files to /etc as part of setup script instead of package install
* Fix warnings and installation of mapnik files
* Add missing python-serial dependency
* Cleanup linking of libturbojpeg
  This should fix compiling on other platforms like x86_32
* Contributors: Christopher Berner

1.1.0 (2016-01-18)
------------------
* Auto-detect camera device path
* Improve video streaming FPS ~2x
* Add support for hardware h264 acceleration on Odroid XU4
* Add FPS counter to camera page
* Contributors: Christopher Berner

1.0.3 (2015-08-23)
------------------
* Fix race freeing background image sink resources
* Fix bad free on older versions of libav

static/index.js
/*
 * Copyright 2012 the original author or authors.
 * See the NOTICE file distributed with this work for additional
 * information regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
angular.module('rospilot', ['ngRoute', 'ngResource'])
.config(function($routeProvider) {
    $routeProvider
    .when("/graphs", {templateUrl:'static/graphs.html', controller:'graphs'})
    .when("/camera", {templateUrl:'static/camera.html', controller:'camera'})
    .when("/settings", {templateUrl:'static/settings.html', controller:'settings'})
    .when("/flight_control", {templateUrl: 'static/flight_control.html', controller: 'position'})
    .otherwise({redirectTo:"/flight_control"});
})
.factory('ROS', function() {
    var url = 'ws://' + window.location.hostname + ':8088';
    var ros = new ROSLIB.Ros({url: url});
    return ros;
})
.factory('$rosservice', function(ROS) {
    return function(service, type) {
        var service = new ROSLIB.Service({
            ros : ROS,
            name : service,
            messageType : type
        });
        return function(args, callback) {
            if (typeof args === 'undefined') {
                args = {};
            }
            if (typeof callback === 'undefined') {
                callback = function(result) {};
            }
            service.callService(new ROSLIB.ServiceRequest(args), callback);
        };
    };
})
.factory('$rosparam', function(ROS) {
    return {
        get: function(key, callback) {
            var param = new ROSLIB.Param({ros: ROS, name: key});
            param.get(callback);
        },
        set: function(key, value) {
            var param = new ROSLIB.Param({ros: ROS, name: key});
            param.set(value);
        }
    };
})
.factory('$rostopic', function(ROS) {
    return function(topic, type) {
        return new ROSLIB.Topic({
            ros : ROS,
            name : topic,
            messageType : type
        });
    };
})
.factory('Camera', function ($rosservice) {
    return {
        take_picture: $rosservice('/take_picture', 'std_srvs/Empty'),
        start_recording: $rosservice('/start_record', 'std_srvs/Empty'),
        stop_recording: $rosservice('/stop_record', 'std_srvs/Empty')
    };
})
.factory('Media', function ($resource) {
      return $resource('api/media');
})
.factory('Status', function ($rostopic, $rosservice) {
      return {
          subscribe: function(callback) {
              $rostopic('/basic_status', 'rospilot/BasicStatus').subscribe(callback);
          },
          set: $rosservice('/set_mode', 'rospilot/SetBasicMode')
      };
})
.factory('Position', function ($rostopic) {
      return $rostopic('/gpsraw', 'rospilot/GPSRaw');
})
.factory('Attitude', function ($rostopic) {
      return $rostopic('/attitude', 'rospilot/Attitude');
})
.factory('RCState', function ($rostopic) {
      return $rostopic('/rcstate', 'rospilot/RCState');
})
.factory('IMU', function ($rostopic) {
      return $rostopic('/imuraw', 'rospilot/IMURaw');
})
.factory('Waypoints', function ($rostopic, $rosservice) {
      return {
          subscribe: function(callback) {
              $rostopic('/waypoints', 'rospilot/Waypoints').subscribe(callback);
          },
          set: $rosservice('/set_waypoints', 'rospilot/SetWaypoints')
      };
})
.controller('rospilot', function ($scope, $route) {
    $scope.$route = $route;
})
.controller('settings', function ($scope, $rosparam, $rosservice) {
    var globService = $rosservice('/glob', 'rospilot/Glob');
    $scope.selected_video_device = '';
    $scope.video_devices = [];
    $rosparam.get('/camera/video_device',
        function(value) {
            $scope.selected_video_device = value;
            // XXX: This shouldn't be necessary
            $scope.$apply();
        }
    );
    globService({pattern: '/dev/video*'}, function(result) {
        $scope.video_devices = result.paths.sort();
    });

    $scope.$watch('selected_video_device', function(new_device) {
        if (new_device) {
            $rosparam.set('/camera/video_device', new_device);
        }
    });
})
.controller('waypoints', function ($scope, $timeout, Waypoints) {
  $scope.waypoints = [];
  $scope.come_here = function() {
      navigator.geolocation.getCurrentPosition(function(location){
          var waypoints = {
              waypoints: [{
              'latitude': location.coords.latitude,
              'longitude': location.coords.longitude,
              'altitude': 5.0
          }]};
          Waypoints.set(waypoints);
      });
  };
  Waypoints.subscribe(function(waypoints) {
      $scope.waypoints = waypoints.waypoints;
      // XXX: This shouldn't be necessary
      $scope.$apply();
  });
})
.controller('rcstate', function ($scope, RCState) {
  $scope.data = {'channel': []};
  RCState.subscribe(function(rcstate) {
      $scope.data = rcstate;
  });
})
.controller('imu', function ($scope, IMU) {
  $scope.data = {
      'gyro': {'x': 0, 'y': 0, 'z': 0},
      'accel': {'x': 0, 'y': 0, 'z': 0},
      'mag': {'x': 0, 'y': 0, 'z': 0}
  };
  IMU.subscribe(function(data) {
      $scope.data = data;
      if ($('#accel_z_chart').length > 0) {
        var series = $('#accel_z_chart').highcharts().series[0];
        var x = (new Date()).getTime();
        series.addPoint([x, data.accel.z], true, true);
      }
  });
})
.controller('status', function ($scope, $timeout, Status) {
  $scope.data = {armed: false};
  $scope.arm = function() {
      Status.set({
          armed: true
      });
  };
  $scope.disarm = function() {
      Status.set({
          armed: false
      });
  };
  Status.subscribe(function(status) {
      $scope.data = status;
  });
})
.controller('position', function ($scope, $timeout, Position, Waypoints) {
  $scope.waypoint_data = {'waypoints': []};
  // If we don't have an internet connection, Google Maps won't have loaded
  if (typeof(google) != 'undefined' && document.getElementById('map-canvas')) {
    $scope.map = new GMaps({
        div: '#map-canvas',
        lat: 37.77,
        lng: -122.4,
        zoom: 17,
        mapType: 'satellite'
    });

    $scope.map.addMarker({
        lat: 37.77,
        lng: -122.4,
        title: 'Multicopter',
        icon: '/static/red-marker.png'
    });
    $scope.marker = $scope.map.markers[0]
    $scope.map.addMarker({
        lat: 0,
        lng: 0,
        title: 'Waypoint',
        icon: '/static/green-marker.png'
    });
    $scope.waypoint = $scope.map.markers[1];

    $scope.map.setContextMenu({
        control: 'map',
        options: [
            {
                title: 'Set Waypoint',
                name: 'set_waypoint',
                action: function(e) {
                    var waypoints = {
                        waypoints: [{
                        'latitude': e.latLng.lat(),
                        'longitude': e.latLng.lng(),
                        'altitude': 5.0
                    }]};
                    Waypoints.set(waypoints);
                },
            }
        ]
    });
  }

  // GMaps doesn't have touch support, so add our own long press detection
  google.maps.event.addListener($scope.map.map, 'mousedown', function(e) {
      $scope.longpress_promise = $timeout(function() {
        $scope.map.buildContextMenu('map', e)
      }, 1000);
  });
  google.maps.event.addListener($scope.map.map, 'mouseup', function(e) {
      $timeout.cancel($scope.longpress_promise);
  });
  google.maps.event.addListener($scope.map.map, 'dragstart', function(e) {
      $timeout.cancel($scope.longpress_promise);
  });
  
  Waypoints.subscribe(function(data) {
      $scope.waypoint_data = data;
      if (data.waypoints.length > 0) {
          var lat = data.waypoints[0].latitude;
          var lng = data.waypoints[0].longitude;
          var pos = new google.maps.LatLng(lat, lng);
          $scope.waypoint.setPosition(pos);
      }
  });

  Position.subscribe(function(position) {
      $scope.data = position;
      if (typeof($scope.map) != 'undefined') {
        // XXX: This should be moved
        var pos = new google.maps.LatLng(position.latitude,
                                        position.longitude);
        $scope.marker.setPosition(pos);
        $scope.map.setCenter(position.latitude, position.longitude);
      }
  });
})
.controller('attitude', function ($scope, Attitude) {
  var compass = document.getElementById("compass");
  var attitude_svg = document.getElementById("attitude_svg");
  var needle = null;
  var needle_translate = null;
  var roll_gauge = null;
  var roll_gauge_translate = null;
  var roll_needle = null;
  compass.addEventListener('load', function() {
      needle = compass.getSVGDocument().getElementById("needle");
      needle_translate = needle.getAttribute("transform");
  });
  attitude_svg.addEventListener('load', function() {
      roll_needle = attitude_svg.getSVGDocument().getElementById("layer2");
      roll_gauge = attitude_svg.getSVGDocument().getElementById("layer5");
      roll_gauge_translate = roll_gauge.getAttribute("transform");
  });
  $scope.data = {'roll': 0, 'pitch': 0, 'yaw': 0};

  Attitude.subscribe(function(attitude) {
      if (needle != null) {
          var x = needle.getBBox().width / 2.0;
          var y = needle.getBBox().height / 2.0;
          var yaw = attitude.yaw * 180 / Math.PI;
          needle.setAttribute("transform", 
              "rotate(" + -yaw + " " + x + " " + y + ") "
              + needle_translate);
      }

      if (roll_gauge != null) {
          var x = roll_needle.getBBox().width / 2.0;
          var y = roll_needle.getBBox().height / 2.0;
          var roll = -attitude.roll * 180 / Math.PI;
          var pitch = attitude.pitch * roll_gauge.getBBox().height / Math.PI;
          roll_needle.setAttribute("transform", 
              "rotate(" + roll + " " + x + " " + y + ")");
          roll_gauge.setAttribute("transform", 
              "rotate(" + roll + " " + x + " " + y + ") " +
              roll_gauge_translate + " translate(0 " + pitch + ")");
      }

      $scope.data = attitude;
  });
})
.controller('graphs', function($scope) {
  Highcharts.setOptions({
      global: {
          useUTC: false
      }
  });

  var chart;
  $('#accel_z_chart').highcharts({
    chart: {
        type: 'spline',
        animation: false,
        marginRight: 10,
    },
    title: {
        text: 'Accelerometer'
    },
    xAxis: {
        type: 'datetime',
        tickPixelInterval: 150
    },
    yAxis: {
        title: {
            text: 'Value'
        },
        plotLines: [{
            value: 0,
            width: 1,
            color: '#808080'
        }]
    },
    tooltip: {
        formatter: function() {
                return '<b>'+ this.series.name +'</b><br/>'+
                Highcharts.dateFormat('%Y-%m-%d %H:%M:%S', this.x) +'<br/>'+
                Highcharts.numberFormat(this.y, 2);
        }
    },
    legend: {
        enabled: false
    },
    exporting: {
        enabled: false
    },
    series: [{
        name: 'data',
        data: (function() {
            var data = [],
                time = (new Date()).getTime(),
                i;

            for (i = -19; i <= 0; i++) {
                data.push({
                    x: time + i * 1000,
                    y: 0
                });
            }
            return data;
        })()
    }]
  });
})
.controller('camera', function($scope, $timeout, Media, Camera) {
  $scope.media = {'objs': []};

  (function tick() {
      Media.get({}, function(data) {
          if (data.objs.length != $scope.media.objs.length) {
            $scope.media = data;
          }
          $timeout(tick, 1000);
      });
  })();

  $scope.server_name = window.location.hostname;
  $scope.take_picture = Camera.take_picture;

  $scope.recording = false;
  $scope.toggle_recording = function() {
    if ($scope.recording) {
      Camera.stop_recording();
    } else {
      Camera.start_recording();
    }
    $scope.recording = !$scope.recording;
  };
})
.directive('activeClass', function($location) {
    return {
        restrict: 'A',
        link: function(scope, element, attrs, controller) {
            var css = attrs.activeClass;
            var a = element.find('a');
            var href = a.attr('href');
            href = href.substring(1);
            scope.location = $location;
            scope.$watch('location.path()', function(path) {
                if (href === path) {
                    element.addClass(css);
                } else {
                    element.removeClass(css);
                }
            });
        }
    };
});
